File: web/src/services/ElevenLabsService.ts
export class ElevenLabsService {
  private audioContext: AudioContext | null = null;

  async synthesizeSpeech(
    text: string,
    voiceID: string,
    apiKey: string,
    volume: number
  ): Promise<ArrayBuffer> {
    if (!apiKey || !voiceID) {
      throw new Error('API key or Voice ID is missing');
    }

    const url = `https://api.elevenlabs.io/v1/text-to-speech/${voiceID}`;
    const response = await fetch(url, {
      method: 'POST',
      headers: {
        'xi-api-key': apiKey,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        text,
        model_id: 'eleven_monolingual_v1',
        voice_settings: {
          stability: 0.5,
          similarity_boost: 0.75,
        },
      }),
    });

    if (!response.ok) {
      throw new Error(`ElevenLabs API error: ${response.statusText}`);
    }

    return await response.arrayBuffer();
  }

  async playAudio(data: ArrayBuffer, volume: number): Promise<void> {
    try {
      if (!this.audioContext) {
        this.audioContext = new AudioContext();
      }

      const audioBuffer = await this.audioContext.decodeAudioData(data);
      const source = this.audioContext.createBufferSource();
      const gainNode = this.audioContext.createGain();

      source.buffer = audioBuffer;
      gainNode.gain.value = volume;

      source.connect(gainNode);
      gainNode.connect(this.audioContext.destination);

      source.start(0);
    } catch (error) {
      console.error('Error playing audio:', error);
      throw error;
    }
  }

  stopAudio(): void {
    // Web Audio API doesn't have a direct stop method for all sources
    // We'll need to track sources if we want to stop them
    // For now, this is a placeholder
  }

  async previewVoice(
    text: string,
    voiceID: string,
    apiKey: string,
    volume: number
  ): Promise<void> {
    try {
      const data = await this.synthesizeSpeech(text, voiceID, apiKey, volume);
      await this.playAudio(data, volume);
    } catch (error) {
      console.error('Preview error:', error);
      throw error;
    }
  }
}

File: Services/ElevenLabsService.swift
//
//  ElevenLabsService.swift
//  PomodoroVoice
//

import Foundation
import AVFoundation

class ElevenLabsService: ObservableObject {
    private var audioPlayer: AVAudioPlayer?
    
    func synthesizeSpeech(text: String, voiceID: String, apiKey: String, volume: Double, completion: @escaping (Result<Data, Error>) -> Void) {
        guard !apiKey.isEmpty, !voiceID.isEmpty else {
            completion(.failure(NSError(domain: "ElevenLabsService", code: 1, userInfo: [NSLocalizedDescriptionKey: "API key or Voice ID is missing"])))
            return
        }
        
        let url = URL(string: "https://api.elevenlabs.io/v1/text-to-speech/\(voiceID)")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue(apiKey, forHTTPHeaderField: "xi-api-key")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        let requestBody: [String: Any] = [
            "text": text,
            "model_id": "eleven_monolingual_v1",
            "voice_settings": [
                "stability": 0.5,
                "similarity_boost": 0.75
            ]
        ]
        
        request.httpBody = try? JSONSerialization.data(withJSONObject: requestBody)
        
        URLSession.shared.dataTask(with: request) { data, response, error in
            if let error = error {
                completion(.failure(error))
                return
            }
            
            guard let data = data else {
                completion(.failure(NSError(domain: "ElevenLabsService", code: 2, userInfo: [NSLocalizedDescriptionKey: "No data received"])))
                return
            }
            
            completion(.success(data))
        }.resume()
    }
    
    func playAudio(data: Data, volume: Double) {
        do {
            audioPlayer = try AVAudioPlayer(data: data)
            audioPlayer?.volume = Float(volume)
            audioPlayer?.prepareToPlay()
            audioPlayer?.play()
        } catch {
            print("Error playing audio: \(error)")
        }
    }
    
    func stopAudio() {
        audioPlayer?.stop()
        audioPlayer = nil
    }
    
    func previewVoice(text: String, voiceID: String, apiKey: String, volume: Double) {
        synthesizeSpeech(text: text, voiceID: voiceID, apiKey: apiKey, volume: volume) { [weak self] result in
            DispatchQueue.main.async {
                switch result {
                case .success(let data):
                    self?.playAudio(data: data, volume: volume)
                case .failure(let error):
                    print("Preview error: \(error)")
                }
            }
        }
    }
}

